Iterators

adsharma opened the pull request:
Just so that the code lives up to the name of the library.

We still need to write up some documentation describing the internals. But at a high level, the iterator iterates over Item -> Item unlike rocksdb::Iterator which is a string -> string map.

This pull request contains only the LimitIterator.

manukranthk commented on include/iterlib/Iterator.h:
```diff
@@ -0,0 +1,234 @@
+enum class ResultOrder {
+  DEFAULT = 0, // Index order
+  INDEX_ORDER = 0,
+  ID_ORDER,
```
We can drop specific order types here.

yfeldblum commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,229 @@
+  }
+
+  bool operator<(const ItemOptimized& other) const {
+    return std::tie(value(), ts_, id_)
```
Provide a `private` member function (e.g., `tied`). Then this function will look like:

```
return tied() < other.tied();
```

DRYer, less risk of ordering bugs.

yfeldblum commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,229 @@
+//
+// The function performs a projection as well as comparison and deals with
+// corner cases involving some columns not being present in v1 or v2
+int lessThanOrEqual(const Item& v1,
```
Indentation. Formal parameters start on the next line, are indented +4.

yfeldblum commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,229 @@
+            const std::vector<std::string>& columns,
+            bool useTsIdBackup);
+
+inline std::ostream& operator<<(std::ostream& os, const Item& row) {
```
Make this non-inline and move this out of the header. Then the header will need to include only `<iosfwd>`, not `<ostream>` or `<iostream>`.

yfeldblum commented on include/iterlib/LimitIterator.h:
```diff
@@ -0,0 +1,38 @@
+  }
+
+protected:
+ bool doNext() override;
```
Indentation.

yfeldblum commented on src/Item.cpp:
```diff
@@ -0,0 +1,110 @@
+        attrLess = attrEqual ? false : (isDescending ? (attr1 < attr2)
+                                                     : (attr1 > attr2));
+      } catch (const std::exception& ex) {
+        // == does not throw, but < does, protecting against such cases
```
wut

yfeldblum commented on src/Item.cpp:
```diff
@@ -0,0 +1,110 @@
+
+    if (v1.id() != v2.id()) {
+      return v1.id() < v2.id();
+    }
```
Use `std::tie`.

yfeldblum commented on src/Item.cpp:
```diff
@@ -0,0 +1,110 @@
+  return -1;
+}
+
+int lessThanOrEqual(const Item& v1,
```
Parameter indentation.

yfeldblum commented on src/WrappedIterator.cpp:
```diff
@@ -0,0 +1,33 @@
+  if (innerIter_) {
+    return innerIter_->prepare()
+      .then([this]() {
+          // Code to extract any data from inner iter
```
Indentation. Lambda code block is indented +2 compared to outer scope (this is +4).

yfeldblum commented on src/WrappedIterator.cpp:
```diff
@@ -0,0 +1,33 @@
+      .then([this]() {
+          // Code to extract any data from inner iter
+          // Hopefully none in a zero-copy implementation
+        })
```
Closing `})` should not be indented compared to the opening `{(`.

yfeldblum commented on tests/IteratorTest.h:
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
Rename this header....

adsharma replied:
What's wrong with IteratorTest.h? Did you mean something like ExpectIterator.h?

adsharma commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,229 @@
+
+  Item(const Item& other) = default;
+  Item& operator=(const Item& other) = default;
+  Item& operator=(Item&& other) = default;
```
will do.

adsharma commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,229 @@
+  int64_t ts_;
+};
+
+// lessThanOrEqual is a 3-value function:
```
Looks cleaner. Will change.

manukranthk replied:
We can call the it 
Ordering => predicate?

adsharma commented on tests/IteratorTest.cpp:
```diff
@@ -0,0 +1,38 @@
+                 {"int2", 2L}}},
+    {4, 0, unordered_map_t{{"int1", 1L}, {"str1", std::string("banana")},
+                 {"int2", 4L}}},
+  };
```
@manukranthk  suggests clang-format -i on all the code.

adsharma commented on tests/IteratorTest.h:
```diff
@@ -0,0 +1,27 @@
+namespace iterlib {
+
+template <typename T>
+void ExpectIterator(Iterator* it, const std::vector<T>& expected) {
```
https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md

has some documentation on `testing::MatcherInterface`. Will read up.

adsharma commented:
Ran clang-format on all the changed files. Let me know if you still see any style/indentation issues.

adsharma commented on tests/IteratorTest.cpp:
```diff
@@ -0,0 +1,38 @@
+  ExpectIterator(limitIt.get(), expected);
+}
+
+int main(int argc, char* argv[]) {
```
For people installing folly via rpms/debs, TestMain.cpp is not a part of libfolly.so.
It's not clear how to link to it.

yfeldblum replied:
See `folly/test/Makefile.am`:

https://github.com/facebook/folly/blob/v2016.11.14.00/folly/test/Makefile.am#L19-L34

adsharma replied:
make install doesn't seem to install TestMain.cpp in a library of its own. The method above works only for tests in folly itself.

Would like to keep it simple for target users of iterlib for now.

yfeldblum replied:
Ah I guess that is something to be fixed in Folly.

adsharma commented:
Update addresses many review comments so far. Still TODO - custom gtest matcher for ExpectIterator.

yfeldblum commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,205 @@
+//
+// LT  => less than
+// EQ  => equal
+// GT_OR_NOT_COMPARABLE  => uncomparable or greater than
```
Why are `GT` and `NOT_COMPARABLE` put together?

adsharma replied:
Because I wasn't brave enough to make the changes in one step. Will break it into two.

adsharma commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,226 @@
+    return dynamic::operator==(static_cast<const dynamic&>(other));
+  }
+
+  id_t id() const {
```
Docs in Internals.md

adsharma commented:
I think the updated diff addresses most of the comments. Please let me know if I missed anything.

https://paste.ubuntu.com/23497246/

Is how far I got on using EXPECT_THAT. It's blocked by the gtest on my box not being new enough to support MatchResultListener. Ok to defer to another diff?

yfeldblum commented:
Okay to defer using gmock to another PR.

yfeldblum commented:
And okay to defer making `partialCompare` do a real partial compare (4-valued).

adsharma commented:
partialCompare already returns 4 values. Opened #8 for ExpectIterator

manukranthk commented on include/iterlib/Iterators.h:
```diff
@@ -0,0 +1,25 @@
+#include "iterlib/RocksDBIterator.h"
+
+// Most common Iterators
+#include "iterlib/FilterIterator.h"
```
These iterators aren't added yet.

adsharma replied:
Will drop this file.

manukranthk commented on include/iterlib/LimitIterator.h:
```diff
@@ -0,0 +1,31 @@
+
+namespace iterlib {
+
+class LimitIterator : public WrappedIterator {
```
Should we add block comments for the classes to explain the behavior of the specific iterators?

manukranthk replied:
Especially things like
startOffset_

adsharma replied:
I have a spec which defines all these behaviors in more detail. Will work on getting that documentation polished before adding to the repo. I can add a one line description here though.

manukranthk commented on src/Item.cpp:
```diff
@@ -0,0 +1,89 @@
+
+const Item Item::kEmptyItem{kEmptyItemInternal};
+
+PartialOrder partialCompare(const Item& v1, const Item& v2,
```
Might be good to add some documentation explaining a parallel to SQL
`ORDER BY column_name ASC|DESC, column_name ASC|DESC; `

adsharma replied:
I couldn't come up with a SQL analog (only one row here). Added a simpler comment.

manukranthk commented on include/iterlib/Iterator.h:
```diff
@@ -0,0 +1,211 @@
+
+  // whether this iterator has been prepared. Only after the iterator has
+  // been prepared, we can call next() and access all the getters.
+  bool prepared_;
```
Should this be std::atomic ?

adsharma replied:
We'll need a lock to protect the state being prepared anyway.

adsharma commented on include/iterlib/QueryConstants.h:
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+namespace iterlib {
+namespace QueryConstants {
```
Will remove

adsharma commented:
Could you take a look at the updated diff?

manukranthk commented on include/iterlib/FutureIterator.h:
```diff
@@ -0,0 +1,77 @@
+ */
+template <typename T = Item> class FutureIterator : public Iterator {
+ public:
+  explicit FutureIterator(folly::Future<std::vector<T>>&& f,
```
We can generalize the FutureIterator to have a lazy yield style pagination/next methods.

adsharma replied:
The rocksdb iterator in #9 is such a lazy iterator. Sure, we can look into some lazy iterators for testing purposes (or just use rocksdb iterator for now).

manukranthk commented on include/iterlib/Item.h:
```diff
@@ -0,0 +1,202 @@
+    // This is a heuristic. The actual sync depends on what was
+    // modified. Use syncIdTs/syncMap directly depending on the
+    // context.
+    if (id_ == 0) {
```
This should be a self documenting constant.

manukranthk commented on include/iterlib/Iterator.h:
```diff
@@ -0,0 +1,211 @@
+
+  virtual ~Iterator() {}
+
+  static id_t max() { return std::numeric_limits<id_t>::max(); }
```
Are min max functions required in Iterator?

adsharma replied:
Not in this pull request, but some of the upcoming pull requests use them.
